src/mlpack/methods/ann/layer/perametric_relu_impl.hpp

```diff
@@ -0,0 +1,83 @@
+/**
+ * @file perametric_relu_impl.hpp
+ * @author Prasanna Patil
+ *
+ * Definition of PerametricReLU layer first introduced in the,
+ * Kaiming He, Xiangyu Zhang, Shaoqing, Ren Jian Sun,
+ * "Delving Deep into Rectifiers:
+ * Surpassing Human-Level Performance on ImageNet Classification", 2014
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_PERAMETRICRELU_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_PERAMETRICRELU_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "perametric_relu.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType>
+PerametricReLU<InputDataType, OutputDataType>::PerametricReLU(
+    const double user_alpha) : user_alpha(user_alpha)
+{
+  alpha.set_size(1,1);
+  alpha(0) = user_alpha;
+}
+
+template<typename InputDataType, typename OutputDataType>
+void PerametricReLU<InputDataType, OutputDataType>::Reset()
+{
+  //! Set value of alpha to the one given by user.
+  alpha(0) = user_alpha;
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename InputType, typename OutputType>
+void PerametricReLU<InputDataType, OutputDataType>::Forward(
+    const InputType&& input, OutputType&& output)
+{
+  Fn(input, output);
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename DataType>
+void PerametricReLU<InputDataType, OutputDataType>::Backward(
+    const DataType&& input, DataType&& gy, DataType&& g)
+{
+  DataType derivative;
+  Deriv(input, derivative);
+  g = gy % derivative;
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void PerametricReLU<InputDataType, OutputDataType>::Gradient(
+    const arma::Mat<eT>&& input, arma::Mat<eT>&& error,
+    arma::Mat<eT>&& gradient)
+{
+  if (gradient.n_elem == 0) {
+    gradient = arma::zeros<arma::mat>(1, 1);
+  }
+
+  arma::mat zeros = arma::zeros<arma::mat>(input.n_rows, input.n_cols);
+  gradient(0) = arma::accu(error % arma::min(zeros, input)) / input.n_cols;
```

Comment:
Right now you test the layer with an input of size 1x1 but it would be great if you could test it with an input nxm where n is > 1. The JacobianHardTanHLayerTest test case in ann_layer_test.cpp is one test that you could adapt.